64 abstractmeasurementsofast

bbean23 opened the pull request:
- Extract common pieces of MeasurementSofastFringe.py and MeasurementSofastFixed.py to AbstractMeasurementSofast.py
- add method for creating standard fringes to Fringes class
- "measure_point" -> "v_measure_point_facet"
- "optic_screen_dist" -> "dist_optic_screen"
- fix test_sensitive_strings.py
- patch tkinter.Tk() occasionally failing

Follow up to #43 
Precursor to #60

braden6521 commented on opencsp/app/sofast/SofastService.py:
```diff
@@ -263,14 +315,21 @@ def get_exposure(self) -> float | None:
 
     def set_exposure(self, new_exp: float) -> None:
-        """Sets camera exposure time value to the given value"""
+        """Sets camera exposure time value to the given value (seconds)"""
```
This actually varies by camera. For instance, with the Basler cameras, it is micro seconds.

bbean23 replied:
Yeah, that makes sense.

What do you think about using microseconds as a standard unit of time for exposure, and doing the conversion in the image_acquisition class? (see commit https://github.com/sandialabs/OpenCSP/commit/2e74cefb3a824ef97476fe85c6888e860dfd1fea)

braden6521 replied:
Ooh, I like that. Good idea!

braden6521 commented on opencsp/app/sofast/lib/AbstractMeasurementSofast.py:
```diff
@@ -0,0 +1,98 @@
+    """Sofast measurement (fringe, fixed, etc) data class that contains captured images
+    and metadata about the measurement.
+
+    By abstracting out the common parts of Fringe and Fixed measurements, we allow the OpticScreenDistance to be
+    determined once and used for many measurements. Having a common interface also reduces the burden on the programmer
+    to learn how to use each sofast mode.
```
🎉

braden6521 commented on opencsp/app/sofast/lib/DistanceOpticScreen.py:
```diff
@@ -0,0 +1,56 @@
+@dataclass
+class DistanceOpticScreen(h5.HDF5_IO_Abstract):
+    """Represents a distance measurement between the optic and the center of the screen. This measurement is typically
+    achieved by displaying the crosshairs from the SofastGUI and measuring from the center of the optic
+    (measurement_point 0,0,0) to the center of the crosshairs."""
```
Could we mention that this measurement does not necessarily need to be from the center. Also, for off-axis paraboloids, the center of the optic is not (0, 0, 0). Just trying to reduce potential confusion, but generally I like this.

bbean23 replied:
When you say "does not necessarily need to be from the center", you mean the optic, right? The crosshairs at the center of the projection should always be what is used for this measurement?

If so, then I'd like to just double-check that this is better:
```python
"""Represents a distance measurement between the optic and the center of the screen. This measurement is typically
achieved by displaying the crosshairs from the SofastGUI and measuring from the origin of the optic
(measurement_point 0,0,0) to the center of the crosshairs.

The optic's origin is typically the center of the optic for on-axis optics, such as with spherical or flat mirrors
that are symetric around their midpoint."""
```

braden6521 replied:
Yes, I did mean the optic. And I'm not sure what is best, I might be worrying about this too much. All of our examples typically use (0, 0, 0) because it is easy to find without measuring anything, but the measure point can be any point that is inside the optic's perimeter. Perhaps something like: 

```
"""Represents a distance measurement between the optic and the origin of the screen coordinate system.
This measurement is typically achieved by displaying the crosshairs from the SofastGUI and measuring
from a point on the optic (measurement_point x,y,z) to the center of the crosshairs.

The measure point is typically the center of the optic for on-axis optics, such as with spherical or flat
mirrors that are symmetric around their midpoint."""
```

bbean23 replied:
Updated with your suggestion. I trust that you know this better than I do.

There are some things about how this system works that I still don't understand (big surprise, I know). Maybe some day...

braden6521 commented on opencsp/app/sofast/lib/DistanceOpticScreen.py:
```diff
@@ -0,0 +1,56 @@
+        Loads from HDF file
+
+        """
+        groups, file_names_and_shapes = h5.get_groups_and_datasets(file)
```
Love this.

braden6521 commented on opencsp/app/sofast/lib/Fringes.py:
```diff
@@ -25,6 +25,38 @@ def __init__(self, periods_x: list, periods_y: list):
+    def from_num_periods(cls, fringe_periods_x=4, fringe_periods_y=4) -> 'Fringes':
+        """Creates fringes to be displayed during run_measurement().
+
+        The fringes are displayed as sinusoidal grayscale images. A value of 1 means that only a single large sinusoidal
+        will be used. A higher value will display more images with a faster sinusoidal. Therefore, a higher value will
+        result in finer slope resolution, up to the resolving power of the optical setup.
+
+        Note that a different number of periods may be required for x and for y, if the x and y resolutions differ
+        enough.
```
I strive to make documentation like this.

braden6521 commented on opencsp/app/sofast/lib/MeasurementSofastFixed.py:
```diff
@@ -33,20 +34,19 @@ def __init__(
             Optic to screen distance, meters
```
Could you remove v_measure_point facet and dist_optic_facet from the doc string and replace with dist_optic_screen_measure?

braden6521 commented on opencsp/app/sofast/lib/MeasurementSofastFixed.py:
```diff
@@ -33,20 +34,19 @@ def __init__(
-        self.v_measure_point_facet = v_measure_point_facet
-        self.dist_optic_screen = dist_optic_screen
         self.origin = origin
         self.date = date
         self.name = name
```
Do these need to be here (line 45/46) since they're already set in super().__init__()?

bbean23 replied:
They sure don't. Good catch again!

braden6521 commented on opencsp/common/lib/tool/tk_tools.py:
```diff
@@ -0,0 +1,123 @@
+            # first attempt failed, try again
+            return window_class(*vargs, **kwargs)
+        except Exception:
+            # second attempt failed, give the system a second to stabalize and
+            # try a third time
+            lt.warn("Failed to create a tkinter.Tk() window. Trying again (3rd attempt).")
+            import time
+
+            time.sleep(1)
+            return window_class(*vargs, **kwargs)
+
+
+def _window(*vargs, **kwargs):
+    """Initializes and returns a new tkinter.Tk() instance.
+
+    If creating the window fails, tries again (up to two more times).
+
+    Sometimes initializing a tk window fails. But if you try to
+    initialize the window again, it seems to always succeed.
+    When it fails, it is often with an error about not being able to find a
+    file. Something like::
+
+        _tkinter.TclError: Can't find a usable init.tcl in the following directories
+    """
+    try:
+        # try to create a window
+        return tk.Tk(*vargs, **kwargs)
+    except Exception:
+        try:
+            lt.warn("Failed to create a tkinter.Tk() window. Trying again (2nd attempt).")
+            # first attempt failed, try again
+            return tk.Tk(*vargs, **kwargs)
+        except Exception:
+            # second attempt failed, give the system a second to stabalize and
+            # try a third time
+            lt.warn("Failed to create a tkinter.Tk() window. Trying again (3rd attempt).")
+            import time
+
+            time.sleep(1)
+            return tk.Tk(*vargs, **kwargs)
```
How does it work (mostly for my benefit) with having two functions with the same name?
Also, why does it a private function (with an underscore in front)? Should it not just be `def window(...)`?

bbean23 replied:
Yeah, this was a copy-paste mistake. I also found this error while responding to these comments.

You're also right that it shouldn't be private. Fixed!

braden6521 commented on opencsp/app/sofast/lib/DistanceOpticScreen.py:
```diff
@@ -0,0 +1,56 @@
+    achieved by displaying the crosshairs from the SofastGUI and measuring from the center of the optic
+    (measurement_point 0,0,0) to the center of the crosshairs."""
+
+    v_measure_point_facet: Vxyz = field(default_factory=lambda: Vxyz(0.0, 0.0, 0.0))
```
The xyz inputs into Vxyz are actually supposed to be a length 3 tuple. 
```suggestion
    v_measure_point_facet: Vxyz = field(default_factory=lambda: Vxyz((0.0, 0.0, 0.0)))
```

bbean23 replied:
Erg, whoops. Thanks! 🙏

braden6521 commented on opencsp/app/sofast/lib/DistanceOpticScreen.py:
```diff
@@ -0,0 +1,56 @@
+    """ Optic-screen distance, meters. """
+
+    def save_to_hdf(self, file: str, prefix: str) -> None:
+        """
+        Saves to HDF file
+        """
```
Could these doc strings be filled in more to describe the inputs?

bbean23 replied:
Yeah, they really could be haha. Fixed!

braden6521 commented on opencsp/app/sofast/lib/AbstractMeasurementSofast.py:
```diff
@@ -0,0 +1,98 @@
+"""Measurement class for SofastFringe
+"""
+
+from abc import ABC, abstractmethod
+import datetime as dt
+
+import numpy as np
+
+from opencsp.app.sofast.lib.ImageCalibrationAbstract import ImageCalibrationAbstract
+import opencsp.app.sofast.lib.DistanceOpticScreen as sod
+from opencsp.common.lib.geometry.Vxyz import Vxyz
+import opencsp.common.lib.tool.hdf5_tools as h5
```
Could you remove unused imports:

```suggestion
from abc import ABC
import datetime as dt

import opencsp.app.sofast.lib.DistanceOpticScreen as sod
import opencsp.common.lib.tool.hdf5_tools as h5
```

bbean23 replied:
Done! Thanks for the suggested change!

braden6521 commented on opencsp/app/sofast/lib/MeasurementSofastFixed.py:
```diff
@@ -2,23 +2,24 @@
+import opencsp.app.sofast.lib.AbstractMeasurementSofast as ams
+import opencsp.app.sofast.lib.DistanceOpticScreen as osd
 from opencsp.common.lib.geometry.Vxy import Vxy
 from opencsp.common.lib.geometry.Vxyz import Vxyz
```
Another unused import. Vxyz

braden6521 commented on opencsp/app/sofast/lib/MeasurementSofastFringe.py:
```diff
@@ -6,11 +6,13 @@
 from opencsp.app.sofast.lib.ImageCalibrationAbstract import ImageCalibrationAbstract
+import opencsp.app.sofast.lib.AbstractMeasurementSofast as ams
+import opencsp.app.sofast.lib.DistanceOpticScreen as osd
 from opencsp.common.lib.geometry.Vxyz import Vxyz
```
Another unused import

bbean23 commented:
> @bbean23, sorry it took a while for me to get to this. I've added some comments, feel free to let me know if some of my comments/suggestions are unnecessary.

Your review was exceptional as always. I believe I've addressed everything. Note: I did find an additional bug, as documented in #70

braden6521 reviewed (approved):
Looks good! If we want to change that docstring later, then go ahead. But I'm fine merging this now. Nice work!

braden6521 reviewed (approved):
Approved!
